New Checker for exhausted iterator in nested loops

ranadheerg opened the pull request:
<!--
Thank you for submitting a PR to pylint!

To ease the process of reviewing your PR, do make sure to complete the following boxes.

- [ ] Document your change, if it is a non-trivial one.
  - A maintainer might label the issue ``skip-news`` if the change does not need to be in the changelog.
  - Otherwise, create a news fragment with ``towncrier create <IssueNumber>.<type>`` which will be
    included in the changelog. ``<type>`` can be one of the types defined in `./towncrier.toml`.
    If necessary you can write details or offer examples on how the new change is supposed to work.
  - Generating the doc is done with ``tox -e docs``
- [ ] Relate your change to an issue in the tracker if such an issue exists (Refs #1234, Closes #1234)
- [ ] Write comprehensive commit messages and/or a good description of what the PR does.
- [ ] Keep the change small, separate the consensual changes from the opinionated one.
  Don't hesitate to open multiple PRs if the change requires it. If your review is so
  big it requires to actually plan and allocate time to review, it's more likely
  that it's going to go stale.
- [ ] If you used multiple emails or multiple names when contributing, add your mails
      and preferred name in ``script/.contributors_aliases.json``
-->

## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
|     | :bug: Bug fix          |
| ✓   | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Description

<!-- If this PR references an issue without fixing it: -->

The pull request introduces a new checker, looping-through-iterator (W4801), to detect a specific, high-confidence bug pattern: an iterator defined in an outer scope being exhausted by reuse in a nested loop.

This pattern can lead to silent bugs where an inner loop runs completely on the first iteration of an outer loop and then does nothing on all subsequent iterations. Following the maintainers' advice to start small, this checker is intentionally focused on this nested-loop scenario to ensure high accuracy and avoid false positives.

Refs #2996 

The checker specifically looks for an iterator being consumed in a loop that is nested inside at least one other loop. It correctly handles cases where the iterator is re-defined inside the outer loop (shadowing).

It does not yet detect sequential reuse of an iterator in non-nested loops. This could be a potential enhancement in a future PR.

Running this checker locally against the primer projects successfully identified 4 instances of this bug pattern in established repositories, including Django, music21, and Black, demonstrating its real-world utility.

<!-- If this PR fixes an issue, use the following to automatically close when we merge: -->

Closes #XXXX

Commit:
updates the example for the `docstring-first-line-empty` warning to remove ambiguity about which empty line triggers the warning.

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
New checker for checking iterator reuse within nested loops.

Handles iterator reassignment, shadowing

out of scope
user defined iterators
control flow statements conditional statements, break, continue within loops

ranadheerg committed:
New checker for checking iterator reuse within nested loops.

Handles iterator reassignment, shadowing

out of scope
user defined iterators
control flow statements conditional statements, break, continue within loops

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Pierre-Sassoulas commented:
Thank you for working on this. Looking at the primer we need to detect when Stopiteration are caught so we do not raise in this case.

jacobtylerwalls commented:
The first example from Django is also a false positive given that the outer loop `return`s unconditionally after the first iteration.

ranadheerg commented:
@Pierre-Sassoulas @jacobtylerwalls  
I am enhancing the checker to detect unconditional exits (return, break, raise), which will fix the false positives in the Django and music21 examples.

A question regarding the other Django add_extra example that uses next()

Should we only suppress warnings for next() when a try...except StopIteration block guarantees an exit? My thinking is yes, because the original issue's scope was to prevent silent bugs, whereas an uncaught StopIteration from next() is a runtime error.

Let me know your thoughts.

ranadheerg committed:
fixed false positives for try except and unconditional returns

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

jacobtylerwalls reviewed (changes requested):
It should be possible to reduce the size of this PR. Let's start with the tests.

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
Merge branch 'main' of https://github.com/ranadheerg/pylint into feature/exhausted_iteretor_new_checker

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
migrating unittests to functional test following existing pattern and removing the unit test file

ranadheerg committed:
Merge branch 'main' of https://github.com/ranadheerg/pylint into feature/exhausted_iteretor_new_checker

ranadheerg committed:
removing previous test file as all tests have been moved to functional tests folder

ranadheerg committed:
updating documentation and changelog

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ranadheerg committed:
updating documentation good.py file

ranadheerg commented on tests/checkers/unittest_iterator_checker.py:
```diff
@@ -0,0 +1,1208 @@
+
+
+class TestRepeatedIteratorLoopChecker(CheckerTestCase):
+    """Tests for RepeatedIteratorLoopChecker."""
```
Hi @jacobtylerwalls, I have updated the checker with comprehensive docstrings including Good and Bad examples to clarify the intent of the message. I've also refined the message 'help' text to explain the stateful nature of iterators. The PR now follows the standard Pylint functional testing pattern, and I've added the necessary Towncrier fragment. Ready for re-review!

ranadheerg committed:
updating change log format as expected

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jacobtylerwalls commented on doc/user_guide/configuration/all-options.rst:
```diff
@@ -8,7 +8,6 @@
 Standard Checkers
 ^^^^^^^^^^^^^^^^^
 
-Remember that only ``tool.pylint`` is required, the section title is not. There are specific notes under each example that cover this.
```
?

jacobtylerwalls reviewed (commented):
What is your view about the false positives in Django and black? The Django example uses nested loops to essentially zip-search in one direction through an iterator. The black example also zip-searches in one direction and raises an error in an else branch.

ranadheerg commented:
Hi @jacobtylerwalls — yes, all three are genuine false positives, each hitting a different gap:

**Ansible (`count`)** — `next(count, None)` is deliberate single-element advancement, not bulk consumption. `visit_call` doesn't distinguish the two. Fix: skip the check when the outer call is `next()`.

**Django (`param_iter`)** — `_has_unconditional_exit` bails immediately when it sees an `if` without an `else`, even if there's a `return` after it that both branches hit. Fix: instead of returning `False`, enqueue two sub-paths (through the `if` body, and skipping it) both continuing with the remaining statements — same logic already used for `if/else`.

**Black (`it`)** — This is a `for/else: raise` pattern, a deliberate fail-fast cursor. The author explicitly handles exhaustion by raising, so this isn't a silent bug at all. The checker doesn't look at `for`-`orelse` today. Fix: suppress if the inner `for` loop's `orelse` contains an unconditional exit, and/or if the loop body has a reachable `break` (cursor pattern).

Does that match your read? If so I'll implement — happy to keep them in separate commits if that's easier to review.

ranadheerg committed:
fixing false positives and errors

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ranadheerg committed:
Merge branch 'pylint-dev:main' into main

ranadheerg committed:
Merge branch 'feature/exhausted_iteretor_new_checker' of https://github.com/ranadheerg/pylint into feature/exhausted_iteretor_new_checker

ranadheerg committed:
Merge branch 'main' of https://github.com/ranadheerg/pylint into feature/exhausted_iteretor_new_checker

ranadheerg committed:
fixing false positives and errors

ranadheerg committed:
adding more concrete examples for documentation and fixing linting and precheck issues

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ranadheerg committed:
fixing spelling in documentation

ranadheerg committed:
Merge branch 'pylint-dev:main' into feature/exhausted_iteretor_new_checker

Julfried committed:
Fix augmented sys.path handling for parallel namespace-package checks (#10929)


Co-authored-by: Pierre Sassoulas <pierre.sassoulas@gmail.com>

dependabot[bot] committed:
Bump sigstore/gh-action-sigstore-python from 3.2.0 to 3.3.0 (#10944)

Bumps [sigstore/gh-action-sigstore-python](https://github.com/sigstore/gh-action-sigstore-python) from 3.2.0 to 3.3.0.
- [Release notes](https://github.com/sigstore/gh-action-sigstore-python/releases)
- [Changelog](https://github.com/sigstore/gh-action-sigstore-python/blob/main/CHANGELOG.md)
- [Commits](https://github.com/sigstore/gh-action-sigstore-python/compare/v3.2.0...v3.3.0)

---
updated-dependencies:
- dependency-name: sigstore/gh-action-sigstore-python
  dependency-version: 3.3.0
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

dependabot[bot] committed:
Bump codecov/codecov-action from 5 to 6 (#10943)

Bumps [codecov/codecov-action](https://github.com/codecov/codecov-action) from 5 to 6.
- [Release notes](https://github.com/codecov/codecov-action/releases)
- [Changelog](https://github.com/codecov/codecov-action/blob/main/CHANGELOG.md)
- [Commits](https://github.com/codecov/codecov-action/compare/v5...v6)

---
updated-dependencies:
- dependency-name: codecov/codecov-action
  dependency-version: '6'
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

Pierre-Sassoulas committed:
[test] Add primer test cases for suboptimal diff scenarios (#10946)

Document current behavior where line-only moves, message text changes,
type changes, and useless-suppression pairs all show as full remove+add
in the primer comment. These cases will help validate future smarter
diff logic.

Pierre-Sassoulas committed:
[doc] Upgrade Read the Docs build image to Ubuntu 24.04/Python 3.13 (#10945)

Ubuntu 20.04 is being deprecated by Read the Docs (June 2026), might as well upgrade this one to a more modern Ubuntu and a faster python.

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#10947)

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.15.7 → v0.15.8](https://github.com/astral-sh/ruff-pre-commit/compare/v0.15.7...v0.15.8)

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

Pierre-Sassoulas committed:
[primer] Add blank line after <details> tag for proper markdown rendering (#10948)

GitHub markdown requires a blank line after <details> for the content
inside to be rendered as markdown rather than plain text.

Pierre-Sassoulas committed:
[primer] Separate Comparator data loading from comparison logic

Constructor now takes pre-loaded PackageMessages dicts, making it
directly testable with in-memory data.  I/O and batch merging move to
a from_json() factory method; _load_json() becomes module-level.

Pierre-Sassoulas committed:
[primer] Switch from JSONReporter to JSON2Reporter

The old JSONReporter is deprecated; use JSON2Reporter which provides
confidence, absolutePath, and camelCase messageId fields.  Update all
primer test fixtures accordingly.

Pierre-Sassoulas committed:
[primer] Use realistic package structure in both_empty fixture

Replace bare {} with a package containing an empty messages list,
which is what the primer actually produces.

Pierre-Sassoulas committed:
[primer] Add multi_package test fixture

Test comparator output across two packages with different change
categories: astroid has a changed diagnostic, astropy has a new
false positive and a removed message.

mbyrnepr2 committed:
Bump the upper bound for `astroid` to fix an editable install error. (#10954)

dependabot[bot] committed:
Bump pylint-dev/backport from 2.1.7 to 2.1.8 (#10956)

Bumps [pylint-dev/backport](https://github.com/pylint-dev/backport) from 2.1.7 to 2.1.8.
- [Release notes](https://github.com/pylint-dev/backport/releases)
- [Commits](https://github.com/pylint-dev/backport/compare/35f69bb3064974a24ba7da019038cd616f09e66d...3adad52e5a8ba2ea731c32037442d507eed68b4c)

---
updated-dependencies:
- dependency-name: pylint-dev/backport
  dependency-version: 2.1.8
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] pre-commit autoupdate (#10957)

updates:
- [github.com/astral-sh/ruff-pre-commit: v0.15.8 → v0.15.9](https://github.com/astral-sh/ruff-pre-commit/compare/v0.15.8...v0.15.9)
- [github.com/pre-commit/mirrors-mypy: v1.19.1 → v1.20.0](https://github.com/pre-commit/mirrors-mypy/compare/v1.19.1...v1.20.0)
- [github.com/tox-dev/pyproject-fmt: v2.20.0 → v2.21.0](https://github.com/tox-dev/pyproject-fmt/compare/v2.20.0...v2.21.0)

Co-authored-by: Pierre Sassoulas <pierre.sassoulas@gmail.com>

mvanhorn committed:
feat(dict-init-mutate): include suggested dict literal in message (#10935)

The dict-init-mutate message now shows the dictionary literal
that the user could write, truncated if the dict would be too long.

Closes #7819

Co-authored-by: Matt Van Horn <455140+mvanhorn@users.noreply.github.com>
Co-authored-by: Pierre Sassoulas <pierre.sassoulas@gmail.com>

ranadheerg commented:
@jacobtylerwalls @Pierre-Sassoulas 

Just a quick ping to see if we can get this moving. To rephrase the current state:

I reduced the scope to just nested for loops (removing next() and while loop checks) as discussed.

This successfully eliminated the false positives we saw in the primer run for Django and Black.

The for-else bug is fixed, docs are updated, and all tests/CI checks are completely green.

The PR is currently blocked by the older "Changes requested" review. Could you take a fresh look when you have a minute and let me know if there's anything else needed to clear the block and get this merged?

Also let me know if we want to expand it for more iterator producing functions, I have added the basic functions 

map, filter, zip, iter and reversed

Pierre-Sassoulas commented:
Hey @ranadheerg thank you for your work on this. Do you mind rebasing on main ? The primer seems off.

Pierre-Sassoulas commented on doc/data/messages/l/looping-through-iterator/bad.py:
```diff
@@ -0,0 +1,21 @@
+# 1. Generator expressions are exhausted after one use
```
I think we can keep only one example for the doc, it's to illustrate the issue and the fix not to show that the checker handles multiple situation :)

Pierre-Sassoulas commented on pylint/checkers/looping_iterator_checker.py:
```diff
@@ -0,0 +1,193 @@
+        self._scope_stack: list[dict[str, nodes.NodeNG | str]] = []
+
+    # --- Scope Management ---
+
```
The  @utils.only_required_for_messages("looping-through-iterator") is missing for some visitor function, right ?

Pierre-Sassoulas commented on pylint/checkers/looping_iterator_checker.py:
```diff
@@ -0,0 +1,193 @@
+if TYPE_CHECKING:
+    from pylint.lint import PyLinter
+
+DefinitionType = nodes.NodeNG | str
```
Unused ?

```suggestion
```

Pierre-Sassoulas commented on pylint/checkers/looping_iterator_checker.py:
```diff
@@ -0,0 +1,193 @@
+        "builtins.zip",
+        "builtins.iter",
+        "builtins.reversed",
+    }
```
```suggestion
    KNOWN_ITERATOR_PRODUCING_FUNCTION_QNAMES: set[str] = {
        "builtins.map",
        "builtins.filter",
        "builtins.zip",
        "builtins.iter",
        "builtins.reversed",
    }
```

Pierre-Sassoulas commented on pylint/checkers/looping_iterator_checker.py:
```diff
@@ -0,0 +1,193 @@
+
+    @utils.only_required_for_messages("looping-through-iterator")
+    def visit_call(self, node: nodes.Call) -> None:
+        if isinstance(node.func, nodes.Name) and node.func.name == "next":
```
```suggestion
        if isinstance(node.func, nodes.Name) and node.func.name == "next":
            # When the user use 'next' it's easy to raise a false positive 
            # and the user usually know what they do so the primer looked 
            # especially bad.
```

Pierre-Sassoulas reviewed (changes requested):
This look great overall, I have a few minor comments.

ranadheerg committed:
Update pylint/checkers/looping_iterator_checker.py

Adding note for next calls

Co-authored-by: Pierre Sassoulas <pierre.sassoulas@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ranadheerg committed:
adding minor changes

ranadheerg committed:
adding minor changes

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

ranadheerg committed:
moving comment to outside of if

ranadheerg committed:
Merge branch 'pylint-dev:main' into feature/exhausted_iteretor_new_checker

jacobtylerwalls commented on pylint/checkers/looping_iterator_checker.py:
```diff
@@ -0,0 +1,195 @@
+        """Walks up the AST from a node to find the first containing loop."""
+        current: nodes.NodeNG | None = node
+        while current:
+            if isinstance(current, nodes.For):
```
Got it.

Can you simplify this and just model it on astroid's [_get_if_statement_ancestor](https://github.com/pylint-dev/astroid/blob/88d20dc6037d9af5283be89bfc9f1a65e93213ad/astroid/filter_statements.py#L42)()?

jacobtylerwalls commented on pylint/checkers/looping_iterator_checker.py:
```diff
@@ -0,0 +1,200 @@
+
+
+def register(linter: PyLinter) -> None:
+    """This required function is called by Pylint to register the checker."""
```
```suggestion
```

Pierre-Sassoulas replied:
Oops, missed that dumb llm comment :sweat: Sorry !

jacobtylerwalls commented on tests/functional/l/looping_iterator.py:
```diff
@@ -0,0 +1,420 @@
+"""Functional tests for looping-through-iterator checker."""
```
This file is missing a test for a third nesting level:

```py
for ...
    for ...
        for ...
```

jacobtylerwalls commented on pylint/checkers/looping_iterator_checker.py:
```diff
@@ -0,0 +1,200 @@
+            return
+
+        inner_loop = ancestor_loops_of_usage[0]
+        outer_loop = ancestor_loops_of_usage[1]
```
What happens when one of the loops we care about is at index 2?

Pierre-Sassoulas committed:
Update pylint/checkers/looping_iterator_checker.py

Co-authored-by: Jacob Walls <jacobtylerwalls@gmail.com>

Pierre-Sassoulas committed:
Address review: simplify ancestor-loop walk, cover 3-level nesting

- Replace the iterative _find_ancestor_loop with _enclosing_for_loops,
  modelled on astroid's _get_if_statement_ancestor node_ancestors() walk,
  collecting every enclosing for loop and stopping at the scope boundary.
- Make the index-2 behaviour explicit: redefinition in any outer loop is
  handled by the definition_loop scan, while the exit guard keys off the
  immediate enclosing loop only.
- Drop the stray LLM comment in the exit guard.
- Decorate leave_for with only_required_for_messages for scope-stack
  consistency with the other callbacks.
- Add functional tests for three-level nesting (warns, grandparent-exit
  still warns, parent-exit safe, grandparent-redefinition safe).
- Reduce the good.py doc example to a single illustration.

ranadheerg committed:
Merge branch 'pylint-dev:main' into feature/exhausted_iteretor_new_checker

Pierre-Sassoulas committed:
Handle arbitrary nesting depth in looping-through-iterator

The previous logic returned 'safe' whenever the iterator's redefinition
loop appeared anywhere among the usage's enclosing loops. That silenced a
real bug at 3+ nesting levels: when the iterator is redefined in an outer
loop (e.g. the grandparent) but a loop *between* the redefinition and the
consumption re-runs the consumption without a refresh, the iterator is
exhausted on every pass after the first.

Compute a redefinition boundary instead and warn when any loop sits
between the consuming loop and the redefinition, at any depth. Two
functional tests that codified the false negative as 'safe' are flipped
to expect the warning.

Pierre-Sassoulas commented on doc/data/messages/l/looping-through-iterator/bad.py:
```diff
@@ -0,0 +1,5 @@
+# 1. Generator expressions are exhausted after one use
```
```suggestion
```

Pierre-Sassoulas committed:
Detect user generators and more iterator-producing builtins

Add an ``is_generator()`` branch so a call to any user-defined generator
function (``def f(): yield ...``) is recognised as producing a one-shot
iterator, alongside the existing builtin list.

Expand KNOWN_ITERATOR_PRODUCING_FUNCTION_QNAMES with ``enumerate`` and the
finite one-shot ``itertools`` producers (chain, islice, groupby, starmap,
accumulate, product, permutations, combinations, compress, dropwhile,
takewhile, filterfalse, zip_longest, pairwise, batched). Deliberately
excluded: ``tee`` (returns a tuple of iterators), ``count``/``cycle``
(infinite, so the exhaustion pattern does not apply) and re-iterable dict
views.

Pierre-Sassoulas commented on doc/data/messages/l/looping-through-iterator/good.py:
```diff
@@ -0,0 +1,6 @@
+# Materialize the iterator into a collection so it can be reused
```
```suggestion
```
